Makefile

```diff
@@ -16,8 +16,41 @@ protos/cacher/cacher.pb.go: protos/cacher/cacher.proto
 	go generate ./...
 	goimports -w $@
 
-test:
-	go test -race -coverprofile=coverage.txt -covermode=atomic ${TEST_ARGS} ./...
+test: lint test-only
+
+# NOTE: -race requires CGO_ENABLED=1
+test-only:
+	CGO_ENABLED=1 go test -race -coverprofile=coverage.txt -covermode=atomic ${TEST_ARGS} ./...
 
 run: ${binaries}
 	docker-compose up --build server
+
+GOLINT_VERSION ?= v1.41.1
+HADOLINT_VERSION ?= v2.6.1
+SHELLCHECK_VERSION ?= v0.7.2
+OS := $(shell uname)
+LOWER_OS  = $(shell echo $OS | tr '[:upper:]' '[:lower:]')
+ARCH := $(shell uname -m)
+
+# TODO(tstromberg): Change hadolint/shell checks to be fatal once issues can be addressed
+lint: out/linters/shellcheck-$(SHELLCHECK_VERSION)/shellcheck out/linters/hadolint-$(HADOLINT_VERSION) out/linters/golangci-lint-$(GOLINT_VERSION)
+	out/linters/golangci-lint-$(GOLINT_VERSION) run
+	out/linters/shellcheck-$(SHELLCHECK_VERSION)/shellcheck $(shell find . -name "*.sh") || true
+	out/linters/hadolint-$(HADOLINT_VERSION) $(shell find . -name "*Dockerfile") || true
+
+# Shell script linter
+out/linters/shellcheck-$(SHELLCHECK_VERSION)/shellcheck:
+	mkdir -p out/linters
+	curl -sfL https://github.com/koalaman/shellcheck/releases/download/v0.7.2/shellcheck-$(SHELLCHECK_VERSION).$(OS).$(ARCH).tar.xz | tar -C out/linters -xJf -
+
+# Dockerfile linter
+out/linters/hadolint-$(HADOLINT_VERSION):
+	mkdir -p out/linters
+	curl -sfL https://github.com/hadolint/hadolint/releases/download/v2.6.1/hadolint-$(OS)-$(ARCH) > out/linters/hadolint-$(HADOLINT_VERSION)
+	chmod u+x out/linters/hadolint-$(HADOLINT_VERSION)
+
+# Go linter
+out/linters/golangci-lint-$(GOLINT_VERSION):
+	mkdir -p out/linters
+	curl -sfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b out/linters $(GOLINT_VERSION)
+	mv out/linters/golangci-lint out/linters/golangci-lint-$(GOLINT_VERSION)
```

Comment:
stuff like this is what we use `nix-shell` for, have you tried that out? I'd be happy to pair up to get this to use nix-shell instead of fetching linters like this.

Reply:
> hmm this is a pretty huge PR and it has a few commits that do a lot of things. It'd be great if you could re-work so only one linter is enabled per-commit so. This would help review the effects of each linter being enabled.

Do you mean per-linter (golang, hadolint, shellcheck), or do you mean per golang linter module (the ~46 modules) we enable.

If the former: this PR only introduces changes to our code-base necessary to pass golangci-lint, but does demonstrate

If the latter: I do not intend on splitting this PR into 46 PR's.

Reply:
> > hmm this is a pretty huge PR and it has a few commits that do a lot of things. It'd be great if you could re-work so only one linter is enabled per-commit so. This would help review the effects of each linter being enabled.
> 
> Do you mean per-linter (golang, hadolint, shellcheck), or do you mean per golang linter module (the ~46 modules) we enable.

Both.

> 
> If the former: this PR only introduces changes to our code-base necessary to pass golangci-lint, but does demonstrate
> 
> If the latter: I do not intend on splitting this PR into 46 PR's.

Definitely not 46 PRs! I was looking for 46 (probably less) commits. I'd wager most linter modules have no affect to the code base so those could be enabled right away. Then one commit per linter module so we see what it actually does. We'd be able to go through them in this one PR.

Reply:
I have not tried it, but so long as `make lint` works and installs all of its dependencies, I do not really care what tool is used underneath, assuming the tool versions are recent.

Reply:
lets get this PR as is and I'll follow up with nixifying instead of fetching like this.

Reply:
OK. I'll say I am a little bit skeptical on whether nix-shell will be an improvement as you will still need to fetch and install nix-shell as part of the Makefile rule.